src/sinks/prometheus.rs

```diff
@@ -205,8 +230,14 @@ impl Sink for PrometheusSink {
                     hist.observe(val);
                 }
             }),
-            Metric::Set { .. } => {
-                trace!("Sets are not supported in Prometheus sink");
+            Metric::Set { name, val } => {
+                // Sets are implemented using promethius integer counters.
+                self.with_set(name, move |&mut (ref mut counter, ref mut set)| {
+                    if set.insert(val) {
```

Comment:
Also I'm concerned about this collection growing indefinitely, if used for high cardinality sets (user_id, ip, etc). For example, Statsd server resets Set counters every flush period (default 10 seconds). Do you think we can implement something like this? I think we can reset on Prometheus scrape, just after serving the request to `/metrics`.

Reply:
> Seems like HashSet::len() could capture the distinct items count, and we will not need to increment any counters. What do you think?

If we use HashSet::len() then HashSet would need to be wrapped in a Mutex to gain read access in fn handle. That would be trading R atomic read and unique count atomic write in the case of a Counter, for a R mutex read and Event::Set count mutex writes in the case of a Mutex. Where Event::Set count >= unique count. So I think that using separate atomic Counter is a good tradeoff.  



> Also I'm concerned about this collection growing indefinitely, if used for high cardinality sets (user_id, ip, etc). For example, Statsd server resets Set counters every flush period (default 10 seconds). Do you think we can implement something like this? I think we can reset on Prometheus scrape, just after serving the request to `/metrics`.

I agree. That is an issue. I am for flush period since it's unknown when Prometheus scrape will happen, if ever. As for implementing flush, it is possible using separate future which yields for the duration and then resets Counter and HashSet (in this case, Mutex would be needed to gain write access to HashSet, so Counter is not needed and we can use HashSet::len() as you suggested), and repeats. Is there something like that implemented in vector, if not I would suggest implementing it separately using a dedicated future, a registry and a trait, but that depends if it is/will be needed in some other places.

I was also thinking if we could relax the uniques requirement for Set, it would be possible to use something like Bloom Filters to permanently limit the growth. Where the size of the filter would be configurable. That would also encourage often scraping, and would enable us to reset on Prometheus scrap, as you suggested. That way, no separate future is required. 

If it would be acceptable to relax it then I am for that option, but I am not certain that it's acceptable, since that would somewhat break consistency with other Sets.

What do you think?

Reply:
> If we use HashSet::len() then HashSet would need to be wrapped in a Mutex to gain read access in fn handle.

I'm not sure if it's the case. It is possible I'm missing something though.

Shouldn't something like this work?
```rust
self.with_set(name, |(gauge, ref mut set)| {
    set.insert(val);
    gauge.set(set.len() as f64);
})
```

>  I am for flush period since it's unknown when Prometheus scrape will happen, if ever.

Completely agree with that argument and argument about being consistent with other sinks implementing Sets aggregation. Let's add flush period!

Regarding you question about code reuse, I think it's fine to keep the new code in the  Prometheus module for now, and extract what could be shared later when necessary.

> it would be possible to use something like Bloom Filters to permanently limit the growth

This is a good idea, however I don't think it's a better default than old good deterministic counting, which our users might expect. We can open a follow up issue about that.

Reply:
I thought you meant not using Gauge/Counter at all. My mistake.
> Shouldn't something like this work?

Yes you are right. But I would avoid calling set/add/.. on Gauge/Counter as much as possible since they are done with atomic operations.

I agree with all of the rest.

Reply:
> I was also thinking if we could relax the uniques requirement for Set, it would be possible to use something like Bloom Filters to permanently limit the growth. Where the size of the filter would be configurable. That would also encourage often scraping, and would enable us to reset on Prometheus scrap, as you suggested. That way, no separate future is required.
> 
> If it would be acceptable to relax it then I am for that option, but I am not certain that it's acceptable, since that would somewhat break consistency with other Sets.

I like this idea, since it means we can get rid of most of the complexity around the flusher. The [hyperloglog](https://docs.rs/hyperloglog/0.0.11/hyperloglog/struct.HyperLogLog.html) data structure seems very well suited to this use case and should also free us from having to keep separate sets and counters synchronized. We can set the error rate quite low by default and I don't believe the probabilistic nature should cause a big problem for anyone.